### PR TITLE
Fix casts of `ScalarValue::Utf8` to `DictionaryArray`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/2873

 # Rationale for this change
Fixes a bug introduced in https://github.com/apache/arrow-datafusion/pull/2819

See https://github.com/apache/arrow-datafusion/issues/2873 (a query that used to work does not anymore). The underlying issue is described in https://github.com/apache/arrow-datafusion/issues/2874

# What changes are included in this PR?
Hack around https://github.com/apache/arrow-datafusion/issues/2874

# Are there any user-facing changes?
Sone queries on dictionaries work again

cc @viirya

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/try_cast.rs [hunk 1]
```diff
@@ -82,6 +82,19 @@ impl PhysicalExpr for TryCastExpr {
                 &array,
                 &self.cast_type,
             )?)),
+            ColumnarValue::Scalar(scalar)
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/try_cast.rs:85]: This is the actual code change -- the rest of the PR is tests

Comment (tustvold) [reply datafusion/physical-expr/src/expressions/try_cast.rs:85]: Won't this prevent this from using the scalar comparison kernels, effectively reversing #2808

**Review by tustvold — APPROVED** [review]
I'm not a massive fan of this, as I think it will basically prevent evaluating scalar predicates directly against dictionaries, but I'm approving as it is blocking upgrades.

I personally would prefer we carry the data type on ColumnarValue::Scalar or something...

File: datafusion/physical-expr/src/expressions/try_cast.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-expr/src/expressions/try_cast.rs:85]: 🤔 that is a good point -- perhaps I'll have a go at a proper fix for https://github.com/apache/arrow-datafusion/issues/2874

**Comment by alamb** [comment]
Will convert to draft as I attempt to "do it properly" as in  https://github.com/apache/arrow-datafusion/issues/2874

**Comment by alamb** [comment]
Update: Doing it properly in https://github.com/apache/arrow-datafusion/pull/2891 is showing good promise

**Comment by alamb** [comment]
Closing in favor of https://github.com/apache/arrow-datafusion/pull/2891

**Pull request closed without merging**